include/alpaka/dev/DevOacc.hpp

```diff
@@ -138,15 +135,38 @@ namespace alpaka
             };
         } // namespace detail
     } // namespace oacc
+
     //! The OpenACC device handle.
     class DevOacc
         : public concepts::Implements<ConceptCurrentThreadWaitFor, DevOacc>
         , public concepts::Implements<ConceptDev, DevOacc>
     {
         friend struct traits::GetDevByIdx<PltfOacc>;
 
+        static auto& device(int iDevice)
+        {
+            static std::vector<std::shared_ptr<oacc::detail::DevOaccImpl>> devices;
+            static std::mutex mutex;
+
+            if(!devices.size())
+            {
+                std::lock_guard<std::mutex> lock(mutex);
+                if(!devices.size())
+                    devices.resize(getDevCount<PltfOacc>());
+            }
```

Comment:
I am not sure if you can use double-checked locking if the locking condition does a non-atomic read. But either way, you can avoid it alltogether by just providing an initializer for the static variable:

```suggestion
            static auto devices = []{
                std::vector<std::shared_ptr<oacc::detail::DevOaccImpl>> d;
                d.resize(getDevCount<PltfOacc>());
                return d;
            }();
            static std::mutex mutex;
```

Reply:
> I am not sure if you can use double-checked locking if the locking condition does a non-atomic read.

The non-atomicity of the check+lock is the reason for the second check. So, the first check does not need to be atomic. The second check require cache coherence though to work. Is this what you are considering?

> But either way, you can avoid it alltogether by just providing an initializer for the static variable:

I read, that some compilers do not properly implement the thread-safe initialization of static variables, so I decided to keep the initializer as simple as possible. Anyway, if we go with a complex initializer, why not this:
```suggestion
            static std::vector<std::shared_ptr<oacc::detail::DevOaccImpl>> devices(getDevCount<PltfOacc>());
            static std::mutex mutex;
``
?

Reply:
> The non-atomicity of the check+lock is the reason for the second check. So, the first check does not need to be atomic. The second check require cache coherence though to work. Is this what you are considering?

I know the basics of double-checked locking, but I think you are still missing a memory barrier: https://preshing.com/20130930/double-checked-locking-is-fixed-in-cpp11/. Also you cannot read/write to a non-atomic variable outside locks, and the first read of `device.size()` is definitely outside a lock.

Reply:
> I read, that some compilers do not properly implement the thread-safe initialization of static variables, so I decided to keep the initializer as simple as possible.

Then the compiler is not C++11 conforming. This was the case with old VS versions and those are long gone. I would rely on this behavior.

`static std::vector<std::shared_ptr<oacc::detail::DevOaccImpl>> devices(getDevCount<PltfOacc>());` is btw an even better solution!

Reply:
>  but I think you are still missing a memory barrier
> [...]
> Also you cannot read/write to a non-atomic variable outside locks, and the first read of `device.size()` is definitely outside a lock.

The initialization of `devices`, and therefore the variable behind `devices.size()`, happens inside a lock. The first read is not directly sequenced with the "write" in line 155. In the case a thread B reads old data after thread A has already written, then it will proceed to lock the mutex, checking again at the second read which is sequenced and it will get the correct data then. Regarding the initialization, the first read being a data race does not matter, because that is what the mutex is for.

There is another possible issue though, which I do not know if it can be solved without having every thread lock the mutex unconditionally on the first encounter: `devices.resize()` is as far from atomic as it can be. If it writes `devices.size()` before actually allocating the memory and storing the `devices.data()` pointer (and I think the compiler could just reorder these operations) then that thread B could run into trouble, if it reads the updated value of `devices.size()` and then proceeds to dereference a `nullptr`. I guess `std::call_once` takes care of this issue.

> Then the compiler is not C++11 conforming. This was the case with old VS versions and those are long gone. I would rely on this behavior.

I recall something about VS in this context. Good to know, that this is fixed.